### DISCUSSION TITLE
zh/usage/python/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#8736] [Docs]
# zh/usage/python/

使用YOLOv8 通过对象检测、分割和分类提升您的Python 项目。探索如何轻松加载、训练、验证、预测、导出、跟踪和基准模型。

https://docs.ultralytics.com/zh/usage/python/

**Comment by flyingship19820207** [comment]
there is an error while trying to train yolo8 :RuntimeError: Dataset 'coco8.yaml' error ❌ '_lzma.LZMADecompressor' object has no attribute 'needs_input'.I use python3.11 and the system is ubuntu.How to fix it.

**Reply by pderrenger** [reply]
Hey there! 👋 It seems like you're encountering an issue related to the `_lzma` module while training YOLOv8 with Python 3.11 on Ubuntu. This could be due to a compatibility issue with the LZMA library in Python 3.11.

A quick workaround could be to try downgrading to Python 3.8 or 3.9, as these versions are known to work well with YOLOv8 and its dependencies. You can manage multiple Python versions using a tool like `pyenv`.

Here's how you can install and use `pyenv` to switch to Python 3.8:

```bash
# Install pyenv
curl https://pyenv.run | bash

# Add pyenv to path (you might want to add this to your .bashrc or .zshrc)
export PATH="$HOME/.pyenv/bin:$PATH"
eval "$(pyenv init --path)"
eval "$(pyenv virtualenv-init -)"

# Install Python 3.8
pyenv install 3.8.10

# Set the local (current directory) Python version to 3.8.10
pyenv local 3.8.10

# Verify the Python version
python --version
```

After switching to Python 3.8, try running your training command again. If the issue persists or you have any more questions, feel free to reach out! 😊

**Reply by flyingship19820207** [reply]
I comment these code below:
#if zipfile.is_zipfile(file) or is_tarfile(file):
 #       new_dir = safe_download(file, dir=DATASETS_DIR, unzip=True, delete=False)
  #      file = find_dataset_yaml(DATASETS_DIR / new_dir)
 #       extract_dir, autodownload = file.parent, False
in file /usr/local/lib/python3.11/site-packages/ultralytics/data/utils.py,then run again, it seems works.

**Reply by pderrenger** [reply]
Hey there! 👋 It looks like you've found a workaround by commenting out some code related to dataset extraction in `utils.py`. While this might work temporarily, it's generally not recommended to modify library code directly, as it could lead to unexpected behavior or issues with future updates.

The error you encountered seems related to the `_lzma` module in Python 3.11. This could be a compatibility issue with the Python version and the LZMA library. A safer approach might be to ensure all your dependencies are correctly installed and compatible with Python 3.11. You could try reinstalling the `lzma` module or using a Python environment with a version known to work well with YOLOv8, such as Python 3.8 or 3.9.

If you're set on using Python 3.11, you might also want to check if there's an open issue on the [Ultralytics YOLOv8 GitHub repository](https://github.com/ultralytics/yolov8) regarding this, or consider opening a new issue to bring this to the developers' attention. They might offer a more permanent fix or advice on handling this compatibility issue.

In the meantime, if you need to keep working, using a virtual environment with a different Python version could be a quick fix:

```bash
# Create a virtual environment with Python 3.8
python3.8 -m venv yolov8-env

# Activate the virtual environment
source yolov8-env/bin/activate

# Install YOLOv8 and other dependencies
pip install ultralytics
```

Hope this helps, and happy coding! 😊

**Comment by flyingship19820207** [comment]
The code is simple, for it is the same with your  sample:

from ultralytics import YOLO

# Create a new YOLO model from scratch
model = YOLO('yolov8n.yaml')

# Load a pretrained YOLO model (recommended for training)
model = YOLO('yolov8n.pt')

# Train the model using the 'coco128.yaml' dataset for 3 epochs
results = model.train(data='coco128.yaml', epochs=3)

# Evaluate the model's performance on the validation set
results = model.val()

# Perform object detection on an image using the model
results = model('https://ultralytics.com/images/bus.jpg')

# Export the model to ONNX format
success = model.export(format='onnx')


Below is the error details:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/ultralytics/engine/trainer.py", line 133, in __init__
    self.data = check_det_dataset(self.args.data)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/ultralytics/data/utils.py", line 271, in check_det_dataset
    if zipfile.is_zipfile(file) or is_tarfile(file):
                                   ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/tarfile.py", line 2791, in is_tarfile
    t = open(name)
        ^^^^^^^^^^
  File "/usr/local/lib/python3.11/tarfile.py", line 1804, in open
    return func(name, "r", fileobj, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/tarfile.py", line 1933, in xzopen
    t = cls.taropen(name, mode, fileobj, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/tarfile.py", line 1854, in taropen
    return cls(name, mode, fileobj, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/tarfile.py", line 1714, in __init__
    self.firstmember = self.next()
                       ^^^^^^^^^^^
  File "/usr/local/lib/python3.11/tarfile.py", line 2631, in next
    raise e
  File "/usr/local/lib/python3.11/tarfile.py", line 2604, in next
    tarinfo = self.tarinfo.fromtarfile(self)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/tarfile.py", line 1292, in fromtarfile
    buf = tarfile.fileobj.read(BLOCKSIZE)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/lzma.py", line 207, in read
    return self._buffer.read(size)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/_compression.py", line 68, in readinto
    data = self.read(len(byte_view))
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/_compression.py", line 96, in read
    if self._decompressor.needs_input:
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: '_lzma.LZMADecompressor' object has no attribute 'needs_input'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/root/chenyong/1_yolo/1_test/yolo_test.py", line 10, in <module>
    results = model.train(data='coco128.yaml', epochs=3)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/ultralytics/engine/model.py", line 635, in train
    self.trainer = (trainer or self._smart_load("trainer"))(overrides=args, _callbacks=self.callbacks)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/ultralytics/engine/trainer.py", line 137, in __init__
    raise RuntimeError(emojis(f"Dataset '{clean_url(self.args.data)}' error ❌ {e}")) from e
RuntimeError: Dataset 'coco128.yaml' error ❌ '_lzma.LZMADecompressor' object has no attribute 'needs_input'

**Reply by pderrenger** [reply]
It seems like you're encountering an issue related to the LZMA compression library, specifically with the `_lzma.LZMADecompressor` object. This could be due to a variety of reasons, including compatibility issues with the Python version you're using or the specific dataset file format.

As a first step, I recommend ensuring that your Python environment is up-to-date and that you're using a compatible version of the Ultralytics YOLO package. Additionally, verifying the integrity and format of the `coco128.yaml` dataset file might help, as the error suggests there could be an issue with reading the dataset file.

If you're still facing this issue after checking these points, could you please provide more details about your environment, such as the Python and Ultralytics YOLO versions? This information will help us better understand the problem and assist you more effectively. 🛠️

In the meantime, here's a quick check you can perform to ensure your dataset file is accessible and correctly formatted:

```python
import yaml

# Attempt to load the YAML file to check its integrity
with open('coco128.yaml', 'r') as file:
    dataset = yaml.safe_load(file)
    print(dataset)
```

This code snippet will try to load your `coco128.yaml` file and print its contents, which can help identify if there are any issues with the file itself.

**Reply by flyingship19820207** [reply]
Thanks for your reply.
I copy  coco128.yaml from  /usr/local/lib/python3.11/site-packages/ultralytics/cfg/datasets/coco128.yaml to current work dir.And run the code you provided:
import yaml
# Attempt to load the YAML file to check its integrity
with open('coco128.yaml', 'r') as file:
    dataset = yaml.safe_load(file)
    print(dataset)
The result is as below:
{'path': '../datasets/coco128', 'train': 'images/train2017', 'val': 'images/train2017', 'test': None, 'names': {0: 'person', 1: 'bicycle', 2: 'car', 3: 'motorcycle', 4: 'airplane', 5: 'bus', 6: 'train', 7: 'truck', 8: 'boat', 9: 'traffic light', 10: 'fire hydrant', 11: 'stop sign', 12: 'parking meter', 13: 'bench', 14: 'bird', 15: 'cat', 16: 'dog', 17: 'horse', 18: 'sheep', 19: 'cow', 20: 'elephant', 21: 'bear', 22: 'zebra', 23: 'giraffe', 24: 'backpack', 25: 'umbrella', 26: 'handbag', 27: 'tie', 28: 'suitcase', 29: 'frisbee', 30: 'skis', 31: 'snowboard', 32: 'sports ball', 33: 'kite', 34: 'baseball bat', 35: 'baseball glove', 36: 'skateboard', 37: 'surfboard', 38: 'tennis racket', 39: 'bottle', 40: 'wine glass', 41: 'cup', 42: 'fork', 43: 'knife', 44: 'spoon', 45: 'bowl', 46: 'banana', 47: 'apple', 48: 'sandwich', 49: 'orange', 50: 'broccoli', 51: 'carrot', 52: 'hot dog', 53: 'pizza', 54: 'donut', 55: 'cake', 56: 'chair', 57: 'couch', 58: 'potted plant', 59: 'bed', 60: 'dining table', 61: 'toilet', 62: 'tv', 63: 'laptop', 64: 'mouse', 65: 'remote', 66: 'keyboard', 67: 'cell phone', 68: 'microwave', 69: 'oven', 70: 'toaster', 71: 'sink', 72: 'refrigerator', 73: 'book', 74: 'clock', 75: 'vase', 76: 'scissors', 77: 'teddy bear', 78: 'hair drier', 79: 'toothbrush'}, 'download': 'https://ultralytics.com/assets/coco128.zip'}

**Reply by flyingship19820207** [reply]
python version is  Python 3.11.7 (main, Jan 15 2024, 10:22:22) [GCC 11.4.0] on linux
YOLO version is 8.1.24

**Reply by pderrenger** [reply]
@flyingship19820207 hey there! Thanks for reaching out with your environment details. It looks like you're using Python 3.11.7 and YOLOv8 version 8.1.24, which should be compatible. 

Given the error you encountered, it seems like there might be an issue with the LZMA library in Python 3.11.7 or how the dataset is being processed. As a quick workaround, you might want to try running your code in a virtual environment with an earlier version of Python (e.g., Python 3.8 or 3.9) to see if the issue persists. This can help isolate whether it's a Python version-specific issue.

Here's a quick guide on setting up a virtual environment:

```bash
# Create a virtual environment
python3 -m venv yolov8-env

# Activate the virtual environment
source yolov8-env/bin/activate

# Install a specific version of Python if needed
pip install python==3.8

# Install Ultralytics YOLO
pip install ultralytics
```

After setting up and activating your virtual environment, try running your YOLOv8 code again. If the problem is resolved, it might be worth reporting this as a compatibility issue with Python 3.11.7 on the Ultralytics GitHub page for further investigation. 🕵️‍♂️

Let me know how it goes!

**Reply by flyingship19820207** [reply]
Thanks for your advice, I will try using python3.9 today.

**Reply by flyingship19820207** [reply]
I have tried python3.10 which system-provided by ubuntu 22.4,it works well.It seems the error is caused by many python version in one system.

**Reply by pderrenger** [reply]
Hey there! 🎉 It's great to hear that switching to Python 3.10 resolved the issue for you. It seems like the problem was indeed related to having multiple Python versions in your system. This can sometimes lead to conflicts, especially with dependencies and libraries that are sensitive to the Python version.

For future reference, it's always a good idea to use virtual environments for each project. This way, you can manage dependencies more effectively and avoid version conflicts. Here's a quick way to set up a virtual environment for your projects:

```bash
python3 -m venv myenv
source myenv/bin/activate
```

And then, you can install the specific packages you need within this isolated environment.

If you have any more questions or run into further issues, feel free to reach out. Happy coding with YOLOv8! 🚀

**Comment by yirenyixin** [comment]
Hello, when I was training with the dataset, I found that all the label names I gave were changed to class_X after training. The following is the configuration of my ymal file：
path: E:\BaiduNetdiskDownload\Citypersons
train: images/train
val: images/train
nc: 1
name:
  0: person

After training, use the model to recognize others. The label information displays class_0

**Reply by glenn-jocher** [reply]
Hello! It seems like the class names are defaulting to `class_X` because the model might not be recognizing the class names specified in your YAML file. This can sometimes happen if there's an issue with the formatting of the YAML file or how the classes are being read during training.

To ensure your class names are recognized correctly, please make sure your YAML file is formatted correctly. Here's a corrected version of your YAML configuration:

```yaml
path: E:\BaiduNetdiskDownload\Citypersons
train: images/train
val: images/train
nc: 1
names:
  - person
```

Notice the change from `name:` to `names:` and ensuring that the class names are listed with a dash `-`. This should help the model recognize and use the correct class names during training and inference.

If the issue persists after this adjustment, please double-check the path and file names for typos and ensure that the dataset is being loaded correctly during training. If you need further assistance, feel free to ask! 😊

**Comment by Sheepy1977** [comment]
There are some translation-mistakes in Chinese
"Train" should be "训练",not “火车”
"Val" should be "校验",not "瓦尔"
"Track“ should be "跟踪“,not "轨道"

**Reply by pderrenger** [reply]
@Sheepy1977 thank you for pointing out the translation errors. We appreciate your attention to detail. We will update the translations as follows:
- "Train" will be corrected to "训练"
- "Val" will be corrected to "校验"
- "Track" will be corrected to "跟踪"

Your feedback helps us improve the documentation for all users. If you notice any other issues, please let us know.

**Comment by QSHCUPID** [comment]
results = model("https://ultralytics.com/images/bus.jpg")

**Reply by QSHCUPID** [reply]
results = model("https://ultralytics.com/images/bus.jpg")

**Reply by QSHCUPID** [reply]
123

**Reply by glenn-jocher** [reply]
Thank you for your comment. It looks like you posted "123" without additional context. If you have any specific questions or need assistance with the Ultralytics YOLOv8 repository, please let us know how we can help!

**Comment by Rain020** [comment]
Operating system: Win11, Ultralyicts: 8.3.27, YOLO11.
I have developed a desktop application for PC-side processing. I need to release related GPU memory resources after closing the page or completing tasks. However, I haven’t found an effective way to release the memory to prevent GPU memory leaks.
Looking forward to your response.

**Reply by glenn-jocher** [reply]
@Rain020 to release GPU memory after completing tasks with YOLO11, you can try using PyTorch's `torch.cuda.empty_cache()` function to clear unused cached memory. Additionally, ensure that all model and variable references are deleted and use `del` to remove them before calling the cache clearing function. If issues persist, consider checking for any persistent references that might still occupy memory. For more details, refer to the [PyTorch documentation](https://pytorch.org/docs/stable/notes/cuda.html).

**Reply by Rain020** [reply]
Thank you for your response. I need to conduct some experiments to verify whether our solution is effective.

**Reply by glenn-jocher** [reply]
You're welcome! Please let us know if you encounter any issues or have further questions while conducting your experiments. We're here to help with any YOLO11-related inquiries.
